keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,409 @@
+  - Feature gate name: `KubeProxyDrainingTerminatingNodes`
+  - Components depending on the feature gate: Kube-proxy
+
+###### Does enabling the feature change any default behavior?
+
+Yes. For `eTP:Local` services: Kube-proxy currently just returns the amount of
+endpoints the service has running on the node in question. Kube-proxy will now
+return an answer to health check which also reflects its health AND how many
+endpoints the service has running on the node. Kube-proxy will also start
+failing its health check server when the Node Kube-proxy is running on gets
+tainted with `ToBeDeletedByClusterAutoscaler`.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, by resetting the feature gate back.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Behavior will be restored back immediately.
+
+###### Are there any tests for feature enablement/disablement?
+
+- Validate that TCP ingress connections are gracefully terminated when:
+
+  * communicating with endpoints on terminating nodes.  
+
+  * communicating with endpoints not on terminating nodes, but for which the
+    terminating node act as nexthops for traffic.
+
+- Validate that new ingress connections to `eTP:Local` endpoints fail when
+  Kube-proxy on that Node is not healthy.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+This change is localized to Kube-proxy only. On upgrades Kube-proxy will
+restart, so client connectivity is impacted in any case. If ingress controllers
+are running on Nodes which are tainted with `ToBeDeletedByClusterAutoscaler`,
+but which are experiencing delay for draining: then ingress SLAs might be
+impacted, whereby ingress connectivity for new connections has a drop below
+what's accepted. But all ingress controllers should be running on these
+terminating Nodes in that case.
+
+###### What specific metrics should inform a rollback?
+
+N/A
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+Two new metrics `proxy_health_check_200_count`/`proxy_health_check_503_count`
+which will count the amount of reported successful/unsuccessful health check
+invocations. A drop in this metric can then be correlated to impacted ingress
+connectivity, for endpoints running on those nodes.
+
+###### How can an operator determine if the feature is in use by workloads?
+
+- By connecting to service of `type: LoadBalancer` and `eTP:Cluster` through a
+  terminating/tainted Node and validating that any new connections are blocked,
+  and established connections are fine.
+- By connecting to service of `type: LoadBalancer` and `eTP:Local` on a
```

Comment:
I don't think we should modify eTP local behavior, there is also a comment from @thockin about that

> I'm advocating, though less strenuously, that we expand eTP:Local LBHC to include other indication(s) of the node's ability to handle traffic (e.g. DP recency). I am NOT advocating to include unschedulable for this. This could impact the availability of the pods on that node, so we must choose wisely (or choose to ONLY consider endpoints-available).

Reply:
You agree that the current behavior is logically flawed, right?

Specifically:

```
1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
  balancer that it should send traffic to a Node simply because the endpoint is
  scheduled there, even though Kube-proxy might not be healthy and successfully
  managed to write the rules required for actually being able to forward traffic
  to the endpoint.
```

Tim is not saying that we shouldn't modify the eTP:Local behavior, _at all_. He is saying that we should modify it to validate that kube-proxy is also healthy, but just not include the Node indicator for this.

Reply:
commented here https://github.com/kubernetes/enhancements/pull/3837/files#r1100013346 eTP local is a bug and we should not need a KEP to fix it. The KEP is nice written and documents perfectly the eTP Cluster problem with draining nodes

Reply:
> eTP local is a bug and we should not need a KEP to fix it.

I think I agree with it.

Reply:
The KEP does spell out some risks with the fix. And I guess that would be lost on a bug fix....do we care?

> 2. If Kube-proxy is flapping state often between healthy and unhealthy then
>    there's risk of disproportionate impact on `eTP:Local`, given that these
>    services won't have any other traffic option other than the node in question.
> 
> 3. Today: if Kube-proxy is unhealthy but an `eTP:Local` service has a ready
>    endpoint on the Node: then Kube-proxy will indicate to the load balancer that
>    it should send that node traffic. This will not be the case anymore once this
>    change is implemented. Clients relying on this behavior will probably see
>    their connections fail.

Reply:
If it streamlines this KEP, I am fine to either break out the `eTP:Local` case to a simple bug or a different KEP.  It is pretty orthogonal

Reply:
OK, let's split it out and deal with it as a bug